File: src/main/kotlin/org/example/schwimmen/search/common/Initialization.kt
package org.example.schwimmen.search.common

import org.example.schwimmen.model.Konfiguration
import org.example.schwimmen.model.Staffel
import org.example.schwimmen.model.StartBelegung
import org.example.schwimmen.search.StaffelBelegung
import org.example.schwimmen.search.Team

// slow (name-based lookup)
fun initialRandomAssignment(konfiguration: Konfiguration): List<Team> =
    List(konfiguration.anzahlTeams) { generateTeam(it + 1, konfiguration) }

private fun generateTeam(
    teamNumber: Int,
    konfiguration: Konfiguration,
) = Team(
    name = "Team $teamNumber",
    staffelBelegungen = konfiguration.staffeln.map { staffel -> generateStaffelbelegung(staffel, konfiguration) },
    konfiguration = konfiguration,
)

private fun generateStaffelbelegung(
    staffel: Staffel,
    konfiguration: Konfiguration,
) = StaffelBelegung(
    staffel = staffel,
    konfiguration = konfiguration,
    startBelegungen = staffel.startDisziplinPaare.map { disziplinName -> generateStartBelegung(konfiguration, disziplinName) },
)

private fun generateStartBelegung(
    konfiguration: Konfiguration,
    disziplinName: String,
): StartBelegung {
    val disziplinId =
        konfiguration.disziplinNameToId[disziplinName]
            ?: error("Programmierfehler: Disziplin $disziplinName konnte nicht gefunden werden")
    val schwimmerIdsZuDenenEsZeitenGibt = konfiguration.getZeiten(disziplinId)
    // TODO proper type instead of pair?
    val schwimmerId = schwimmerIdsZuDenenEsZeitenGibt.random().schwimmerId
    return StartBelegung(schwimmerId = schwimmerId, disziplinId = disziplinId)
}

File: src/ts/search/common/initialization.ts
import { StaffelBelegung, State, Team } from "../state/state";
import { Konfiguration, StaffelX } from "../../eingabe/konfiguration";

export function initialRandomAssignment(konfiguration: Konfiguration): State {
  return { teams: Array(konfiguration.anzahlTeams).filter(() => generateTeam(konfiguration)) };
}

function generateTeam(konfiguration: Konfiguration): Team {
  return {
    staffelBelegungen: konfiguration.staffeln.map((staffel, staffelId) =>
      generateStaffelBelegung(staffel, staffelId, konfiguration),
    ),
  };
}

function generateStaffelBelegung(staffel: StaffelX, staffelId: number, konfiguration: Konfiguration): StaffelBelegung {
  return {
    staffelId,
    startBelegungen: staffel.disziplinIds.map((diszplinId) => generateStartBelegung(diszplinId, konfiguration)),
  };
}

function generateStartBelegung(disziplinId: number, konfiguration: Konfiguration): number {
  const schwimmerIdsZuDenenEsZeitenGibt = konfiguration.disziplinToSchwimmerZeiten[disziplinId];
  const index = Math.random() * schwimmerIdsZuDenenEsZeitenGibt.length;
  return schwimmerIdsZuDenenEsZeitenGibt[index].schwimmerId;
}
